### PR TITLE
Use Source object to track if a layer is duplicated

### PR DESCRIPTION
**Pull request by mstabrin**
# Description

As suggested by https://github.com/napari/napari/issues/5023 after https://github.com/napari/napari/pull/5010 the Source object could keep track of a potential parent.
Here I propose a draft how this could look like :) 

## Type of change

- [X] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by alisterburt — APPROVED** [review]
this is super clean, love it @mstabrin 🥳 

@brisvag what do you think?

**Review by alisterburt — CHANGES_REQUESTED** [review]
actually, got a bit ahead of myself... a test covering this new functionality should be added prior to merge!

**Comment by brisvag** [comment]
Implementation looks good to me! I would wait a bit to hear other opinions in case we missed something, this is public API and we should be wary with expanding it.

**Comment by alisterburt** [comment]
@napari/core-devs any objections to this?

File: napari/layers/_source.py [hunk 1]
```diff
@@ -22,12 +25,15 @@ class Source(BaseModel):
...
         magicgui widget, if the layer was added via a magicgui widget.
+    parent: Layer, optional
+        parent layer if the layer is a duplicate.
     """
 
     path: Optional[str] = None
     reader_plugin: Optional[str] = None
     sample: Optional[Tuple[str, str]] = None
     widget: Optional[FunctionGui] = None
+    parent: Optional[Layer] = None
```
Comment (Czaki) [inline napari/layers/_source.py:36]: it should be `weakref` to not keep the layer in memory if deleted from interface.

**Comment by mstabrin** [comment]
I think I made the implementation a bit cleaner now by utilizing the `layer_source` context manager in the `_duplicate_layer` function. Additionally, the weakref creation is now part of the pydantic `Source` object `parent` validator.

Furthermore, I added some tests :)

I had to move the import of `_source` in the `Layer` object into the `__init__` function due to circular imports:
https://github.com/mstabrin/napari/blob/4df932ad0b8edd803d4eef1835f8c690ee2ed998/napari/layers/base/base.py#L236

File: napari/layers/_source.py (hunk 1 above)
Comment (mstabrin) [reply napari/layers/_source.py:36]: It is now handled in the `parent` validator

File: napari/layers/_source.py [hunk 2]
```diff
@@ -1,11 +1,14 @@
 from __future__ import annotations
 
+import weakref
 from contextlib import contextmanager
 from contextvars import ContextVar
 from typing import Optional, Tuple
```
Comment (alisterburt) [inline napari/layers/_source.py:6]:
```suggestion
from typing import Optional, Tuple, TYPE_CHECKING
```

File: napari/layers/_source.py [hunk 3]
```diff
@@ -1,11 +1,14 @@
...
 
 from magicgui.widgets import FunctionGui
-from pydantic import BaseModel
+from pydantic import BaseModel, validator
+
+from .base.base import Layer
```
Comment (alisterburt) [inline napari/layers/_source.py:11]:
```suggestion
if TYPE_CHECKING:
    from .base.base import Layer
```

**Comment by alisterburt** [comment]
I added a couple of commits which only import when types are being checked at runtime, could you check whether the circular import is still a problem?

**Comment by mstabrin** [comment]
This unfortunately won't work and that is why I needed to resolve the issue in a different way :) 
We are dealing with a pydantic object, therefore the type hint is not merely a type hint, but is actually used during runtime to verify the type and `Layer` really needs to be imported and be available.

**Comment by alisterburt** [comment]
ooft, I thought this made it so that the actual import still happened when runtime type checking was performed... no matter -> LGTM here!

**Review by alisterburt — APPROVED** [review]

**Comment by alisterburt** [comment]
@Czaki @brisvag if you're happy with this and see it when CI has passed could you merge?

**Review by Czaki — APPROVED** [review]
I'm ok with merging.

File: napari/layers/_source.py [hunk 4]
```diff
@@ -1,11 +1,15 @@
 from __future__ import annotations
 
+import weakref
 from contextlib import contextmanager
 from contextvars import ContextVar
-from typing import Optional, Tuple
+from typing import TYPE_CHECKING, Optional, Tuple
```
Comment (alisterburt) [inline napari/layers/_source.py]:
```suggestion
from typing import Optional, Tuple
```

File: napari/layers/_source.py [hunk 5]
```diff
@@ -1,11 +1,15 @@
...
 
 from magicgui.widgets import FunctionGui
-from pydantic import BaseModel
+from pydantic import BaseModel, validator
+
+if TYPE_CHECKING:
+    from .base.base import Layer
```
Comment (alisterburt) [inline napari/layers/_source.py]:
```suggestion
from .base.base import Layer
```

**Review by brisvag — APPROVED** [review]

**Pull request merged**